x-pack/plugins/enterprise_search/public/applications/app_search/app_logic.ts

```diff
@@ -26,5 +33,93 @@ export const AppLogic = kea<MakeLogicType<IAppValues, IAppActions>>({
         initializeAppData: () => true,
       },
     ],
+    account: [
+      {} as IAccount,
+      {
+        initializeAppData: (_, { appSearch: account }) => account,
+        setOnboardingComplete: (state) => ({
+          ...state,
+          onboardingComplete: true,
+        }),
+      },
+    ],
+    configuredLimits: [
+      {} as IConfiguredLimits,
+      {
+        initializeAppData: (_, { configuredLimits }) => configuredLimits.appSearch,
+      },
+    ],
+    ilmEnabled: [
+      false,
+      {
+        initializeAppData: (_, { ilmEnabled }) => !!ilmEnabled,
+      },
+    ],
+  },
+  selectors: {
+    /**
+     * Massages the `role` data we receive from the Enterprise Search
+     * server into a more convenient format for front-end use
+     */
+    myRole: [
+      (selectors) => [selectors.account],
+      ({ role }) => {
+        if (!role) return {};
```

Comment:
Similar question here. How is it valid to return `{}`? That does not conform to the `IRole` interface as expected 🤔

Reply:
In this case I'm less worried about it than `IConfiguredLimits` since it doesn't have nested objects, and will return undefined (which is generally fine for bools/strings/etc., and will only have issues w/ the can functions).

If you're asking why Typescript isn't throwing a fit, I actually have no idea 😂

Reply:
<img width="1791" alt="dog1" src="https://user-images.githubusercontent.com/1427475/93506676-98558a80-f8ea-11ea-80e0-8f8a40ba39bf.png">
<img width="1791" alt="dog2" src="https://user-images.githubusercontent.com/1427475/93506677-98ee2100-f8ea-11ea-89ab-83dff1ba022a.png">

Reply:
Also for some additional context: `account` (and as such `account.role`) should only ever be empty/undefined under the following real world conditions:

1. Our APIs have gone seriously wrong (in which case likely everything else is also not good)
2. The user does not have access to App Search - in which case they should have access to the App Search plugin period (it will be disabled), unless they're a superuser (in which case they would likely only see the Error Connecting/Setup Guide page, and also presumably have some level of technical understanding/debugging abilities as well).

So the empty objs mostly account for scenarios unlikely to ever happen (i.e.: in our unit tests, which populate empty objs by default but have no greater/real world context), whereas `IRole` applies to the scenario we expect to happen.

Reply:
My example above uses a dog ... but a similar thing could happen where you do something like...

```
const { myRole } = useValues()
// Throws an error because we've returned `{}`
myRole.can()
```

Reply:
I guess my point is that this type definition should give an accurate description of the types you can expect to get back in values, and right now, I don't think it's accurate:

https://github.com/elastic/kibana/blob/e17d61d6133e92a5a6673fc3051066ccef5f91b0/x-pack/plugins/enterprise_search/public/applications/app_search/app_logic.ts#L12-L18

Reply:
You see what I'm saying? According to that interface, myRole will always have an `IRole` object, which implies it will always have a `.can` function, which is not necessarily true.

Reply:
@JasonStoltz sorry, only just refreshed and saw your 2nd comment - I'm not super clear on what the screenshot is conveying/asking, can you clarify?

Reply:
The screenshot was showing that TS will let you do something like you mentioned above in your other comment:

```ts
const someObj = {} as IConfiguredLimits; // will not trigger a typescript error
```

But that it's not a good practice, because that sidesteps type checking and can cause us to encounter run time errors, which is what TS is supposed to avoid.

If we do want to use empty objects as defaults, that's fine, I think we just need to update the values type to reflect that:

```
 export interface IAppValues { 
   hasInitialized: boolean; 
   ilmEnabled: boolean; 
   configuredLimits: {} | IConfiguredLimits; 
   account: {} | IAccount; 
   myRole: {} | IRole; 
 } 
```

Reply:
Does that make sense?

Reply:
I think that's why I'd also opt for `null` as a default, because it's easier to check for than an empty object.

Reply:
Ugh super sorry Jason, GitHub (for some bizarre reason) was showing my new comments but not yours until I hard refreshed the page, apologies if it looked like I was totally ignoring you there.

> I think that's why I'd also opt for null as a default

I don't personally have strong feelings on `null` vs `{}` as a default (cc @byronhulcher on this again), but I will note that we do a lot of destructuring values like `const { something: { else } } = useValues(SomeLogic)` in many places in our codebase, and `null` will cause significant type errors in many places / force us to change to something like:
```js
const { something } = useValues(SomeLogic);
const { else } = something || {};
```
or
```js
const { something: { else } = {} } = useValues(SomeLogic);
```
This is immediately apparently in `const { myRole: { canViewSettings } } = useValues(AppLogic)`, for example.

It's debatable whether that extra dev friction is worth it, and if so, whether it's worth changing that paradigm now vs. later 🤷

Reply:
> If we do want to use empty objects as defaults, that's fine, I think we just need to update the values type to reflect that:

```ts
export interface IAppValues { 
   hasInitialized: boolean; 
   ilmEnabled: boolean; 
   configuredLimits: {} | IConfiguredLimits; 
   account: {} | IAccount; 
   myRole: {} | IRole; 
 }
```

This unfortunately doesn't work the way you'd hope it would. It immediately results in type errors:

<img width="757" alt="" src="https://user-images.githubusercontent.com/549407/93556342-631e5c00-f92d-11ea-9213-ff26c63baaf4.png">

So... you're either coercing `as IRole` for the fallback, or you're coercing it the destructured values as `useValues(AppLogic) as { myRole: IRole }` (also equally useless), or you're writing a shit ton of fallbacks or empty catches.

IMO: That's not the goal of Typescript, and Typescript shouldn't be getting in the way like this. Yes, Typescript is a great _static_ type evaluator that provides good developer documentation and compile-time safety, but I disagree that it's meant to catch _run time_ type errors. It cannot know everything and it especially cannot know real-world conditions. Take the most basic of http fetch examples:

```ts
interface IData {
  hopefullyArray: ['great'];
}

const fetchData = async () => {
  const data: IData = await http.get('/some/api');

  hopefullyArray.forEach((item) => alert(item));
}
```
Everything above will pass Typescript checks. Your interface declares it to be an array and your code wants it to be an array. Is the data you're fetching _actually_ an array? Typescript has absolutely 0 idea and will _not_ prevent type errors from crashing the app. All the most interfaces in the world can do nothing about it, and isn't really meant to.

I know it's not a 1:1 example, but I feel sorta similarly about some of the Kea defaults in this scenario. In the real world, all the data being initialized in AppLogic is 1. available on plugin init, and 2. if it's not - the entire app should be inaccessible (managed by our server-side code which disables plugins/sets error connecting state).

We can either continue running/coding the app assuming that if the app is available, the data is there, or we can assume for some reason that our app is running a weird half-broken state and try to account for an empty obj. IMO, in the case of AppLogic, there's no practical reason to assume the latter. If the app is running, `account`, `myRole`, and `configuredLimits` should all be filled.

One thought is that it might be clearer to initialize AppLogic using Kea props vs. attempting to use defaults/`initializeAppData()`: https://kea.js.org/docs/guide/additional#props. I'm open to giving that a shot if that helps w/ peace of mind.

Reply:
>  compile-time safety, but I disagree that it's really meant to catch run-time type errors

Yes, I 100% agree. The example you gave is great. Compile-time safety ultimate helps you avoid run-time errors. Like if a compiler catches an error like the following at compile-time, you've avoided a run-time error. I think we're on the same page.

```
Dog { 
  bark: () => string;
}
aDog: Dog = { }

// The following would be a run-time error if it wasn't caught in compile-time. 
aDog.bark();
```

Back to this example:
```
 export interface IAppValues { 
   hasInitialized: boolean; 
   ilmEnabled: boolean; 
   configuredLimits: {} | IConfiguredLimits; 
   account: IAccount; 
   myRole: IRole; 
 } 
```

According to this interface, myRole is always an IRole.

so myRole should always have a `can` function. 

However, myRole is `{}` until our API call returns from the server.

So if you are in a component and you do the following:

```
const { myRole} = useValues(AppLogic);
myRole.can()
```

Your code will error out.

That sort of error could have been caught at compile time, which forces you to handle that scenario, by using TS.

Any of the following would probably have that effect:

```
 export interface IAppValues { 
   myRole: Partial<IRole>;
   myRole: {} | IRole; 
   myRole: null | IRole; 
 } 
```

Even if it is perhaps not a problem for AppLogic, as you say, this is a problem we'll almost certainly have elsewhere, so I think I'm expressing concern for the pattern as a whole, not just here specifically.

This Interface is the documentation of our logic values for all of the world to see. Any outsider not familiar with the nuances of our code would see that and say "ok, myRole is clearly always an IRole, it says so right there on the Interface".  I think they would be disappointed to find out that when they try to call `myRole.can`, that it errors out. We are not conforming to our own Interface as provided.

This is my 2 cents. Would love to hear from others @scottybollinger @yakhinvadim @byronhulcher

Reply:
> This unfortunately doesn't work the way you'd hope it would. It immediately results in type errors

Kinda mentioned above, I wonder if we could use `Partial`, which marks all keys as optional.

Reply:
I strongly agree that our types should represent the data as it is not the data as we want it to be.  If it is an expected state that this could be `{}` then it should be documented as such.  Yes it means we need to do a lot of checks or else we get typescript errorrs but if we don't do the checks then we will eventually get runtime errors.  I've been using a similar pattern, initializing variables as `null` or `{}`.  It's pedantic, but unambiguous and unsurprising.

Reply:
I've run into issues with `anything | ISomething` in the past, it just works super bizarrely and not like how you'd want it to. That being said it looks like `Partial<>` is working a treat though - no Typescript errors! 🎉  I like that compromise quite a bit, will push it up here.

> Even if it is perhaps not a problem for AppLogic, as you say, this is a problem we'll almost certainly have elsewhere, so I think I'm expressing concern for the pattern as a whole, not just here specifically.

That's definitely a good point - I would hope/want our code that relies on async data to check for it or use optional chaining. 🤔 In practically they rely on `isLoading`/`dataLoading` states, but checks/guards are almost always good to have.

Reply:
Partial fix: 8222a74 🎉 🎉